From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp90: Export ctype<char>::classic_table publicly and implement _Locinfo::_W_Gettnames.

ctype<char>::classic_table, the public form msvcp100 .. msvcp140 export
of the "C" locale's character class table, and _Locinfo::_W_Gettnames
(msvcp110 .. msvcp140), the time names with the wide day and month names
from the CRT's _W_Gettnames, beside _Gettnames. Also maps the ARM names
of msvcp120's unsigned short operator>> and time_get::get (patch 4548).
msvcp120_app and msvcp_win forward.
---
 dlls/msvcp100/msvcp100.spec         |  4 ++--
 dlls/msvcp110/msvcp110.spec         | 10 +++++-----
 dlls/msvcp120/msvcp120.spec         | 14 +++++++-------
 dlls/msvcp120_app/msvcp120_app.spec | 14 +++++++-------
 dlls/msvcp140/msvcp140.spec         | 10 +++++-----
 dlls/msvcp90/locale.c               | 20 ++++++++++++++++++++
 dlls/msvcp_win/msvcp_win.spec       | 10 +++++-----
 7 files changed, 51 insertions(+), 31 deletions(-)

diff --git a/dlls/msvcp100/msvcp100.spec b/dlls/msvcp100/msvcp100.spec
index d5cb700..2d9a96b 100644
--- a/dlls/msvcp100/msvcp100.spec
+++ b/dlls/msvcp100/msvcp100.spec
@@ -1725,8 +1725,8 @@
 @ extern ?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A cin
 @ cdecl -arch=win32 ?classic@locale@std@@SAABV12@XZ() locale_classic
 @ cdecl -arch=win64 ?classic@locale@std@@SAAEBV12@XZ() locale_classic
-@ stub -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ
-@ stub -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ
+@ cdecl -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ() ctype_char_classic_table
+@ cdecl -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ() ctype_char_classic_table
 @ cdecl -arch=arm ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ thiscall -arch=i386 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ cdecl -arch=win64 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
diff --git a/dlls/msvcp110/msvcp110.spec b/dlls/msvcp110/msvcp110.spec
index 861d2b1..4718d18 100644
--- a/dlls/msvcp110/msvcp110.spec
+++ b/dlls/msvcp110/msvcp110.spec
@@ -1824,9 +1824,9 @@
 @ cdecl -arch=arm ?_W_Getmonths@_Locinfo@std@@QBAPBGXZ(ptr) _Locinfo__W_Getmonths
 @ thiscall -arch=i386 ?_W_Getmonths@_Locinfo@std@@QBEPBGXZ(ptr) _Locinfo__W_Getmonths
 @ cdecl -arch=win64 ?_W_Getmonths@_Locinfo@std@@QEBAPEBGXZ(ptr) _Locinfo__W_Getmonths
-@ stub -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ
-@ stub -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ
-@ stub -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ
+@ cdecl -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
+@ thiscall -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
+@ cdecl -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
 @ cdecl -arch=win32 ?_Winerror_map@std@@YAPBDH@Z(long) _Winerror_map_str
 @ cdecl -arch=win64 ?_Winerror_map@std@@YAPEBDH@Z(long) _Winerror_map_str
 @ cdecl ?_XLgamma@std@@YAMM@Z(float) std__XLgamma_float
@@ -1863,8 +1863,8 @@
 @ extern ?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A cin
 @ cdecl -arch=win32 ?classic@locale@std@@SAABV12@XZ() locale_classic
 @ cdecl -arch=win64 ?classic@locale@std@@SAAEBV12@XZ() locale_classic
-@ stub -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ
-@ stub -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ
+@ cdecl -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ() ctype_char_classic_table
+@ cdecl -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ() ctype_char_classic_table
 @ cdecl -arch=arm ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ thiscall -arch=i386 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ cdecl -arch=win64 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
diff --git a/dlls/msvcp120/msvcp120.spec b/dlls/msvcp120/msvcp120.spec
index a9bdd48..27c9458 100644
--- a/dlls/msvcp120/msvcp120.spec
+++ b/dlls/msvcp120/msvcp120.spec
@@ -655,7 +655,7 @@
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAF@Z(ptr ptr) basic_istream_short_read_short
-@ stub -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z
+@ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z(ptr ptr) basic_istream_short_read_ushort
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z(ptr ptr) basic_istream_short_read_ushort
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z(ptr ptr) basic_istream_short_read_ushort
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAH@Z(ptr ptr) basic_istream_short_read_int
@@ -1785,9 +1785,9 @@
 @ cdecl -arch=arm ?_W_Getmonths@_Locinfo@std@@QBAPBGXZ(ptr) _Locinfo__W_Getmonths
 @ thiscall -arch=i386 ?_W_Getmonths@_Locinfo@std@@QBEPBGXZ(ptr) _Locinfo__W_Getmonths
 @ cdecl -arch=win64 ?_W_Getmonths@_Locinfo@std@@QEBAPEBGXZ(ptr) _Locinfo__W_Getmonths
-@ stub -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ
-@ stub -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ
-@ stub -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ
+@ cdecl -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
+@ thiscall -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
+@ cdecl -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
 @ cdecl -arch=win32 ?_Winerror_map@std@@YAPBDH@Z(long) _Winerror_map_str
 @ cdecl -arch=win64 ?_Winerror_map@std@@YAPEBDH@Z(long) _Winerror_map_str
 @ cdecl ?_XLgamma@std@@YAMM@Z(float) std__XLgamma_float
@@ -1823,8 +1823,8 @@
 @ extern ?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A cin
 @ cdecl -arch=win32 ?classic@locale@std@@SAABV12@XZ() locale_classic
 @ cdecl -arch=win64 ?classic@locale@std@@SAAEBV12@XZ() locale_classic
-@ stub -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ
-@ stub -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ
+@ cdecl -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ() ctype_char_classic_table
+@ cdecl -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ() ctype_char_classic_table
 @ cdecl -arch=arm ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ thiscall -arch=i386 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ cdecl -arch=win64 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
@@ -2546,7 +2546,7 @@
 @ cdecl -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
 @ thiscall -arch=i386 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
 @ cdecl -arch=win64 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
-@ stub -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z
+@ cdecl -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z(ptr ptr ptr long ptr long ptr ptr ptr ptr ptr) time_get_wchar_get_fmt
 @ thiscall -arch=i386 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z(ptr ptr ptr long ptr long ptr ptr ptr ptr ptr) time_get_wchar_get_fmt
 @ cdecl -arch=win64 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@PEBG4@Z(ptr ptr ptr ptr ptr ptr ptr ptr ptr) time_get_wchar_get_fmt
 @ cdecl -arch=arm ?get@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
diff --git a/dlls/msvcp120_app/msvcp120_app.spec b/dlls/msvcp120_app/msvcp120_app.spec
index b62fc8a..bd816e8 100644
--- a/dlls/msvcp120_app/msvcp120_app.spec
+++ b/dlls/msvcp120_app/msvcp120_app.spec
@@ -655,7 +655,7 @@
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAF@Z(ptr ptr) msvcp120.??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAF@Z
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAF@Z(ptr ptr) msvcp120.??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAF@Z
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAF@Z(ptr ptr) msvcp120.??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAF@Z
-@ stub -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z
+@ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z(ptr ptr) msvcp120.??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z(ptr ptr) msvcp120.??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z(ptr ptr) msvcp120.??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAH@Z(ptr ptr) msvcp120.??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAH@Z
@@ -1785,9 +1785,9 @@
 @ cdecl -arch=arm ?_W_Getmonths@_Locinfo@std@@QBAPBGXZ(ptr) msvcp120.?_W_Getmonths@_Locinfo@std@@QBAPBGXZ
 @ thiscall -arch=i386 ?_W_Getmonths@_Locinfo@std@@QBEPBGXZ(ptr) msvcp120.?_W_Getmonths@_Locinfo@std@@QBEPBGXZ
 @ cdecl -arch=win64 ?_W_Getmonths@_Locinfo@std@@QEBAPEBGXZ(ptr) msvcp120.?_W_Getmonths@_Locinfo@std@@QEBAPEBGXZ
-@ stub -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ
-@ stub -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ
-@ stub -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ
+@ cdecl -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ(ptr ptr) msvcp120.?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ
+@ thiscall -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ(ptr ptr) msvcp120.?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ
+@ cdecl -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ(ptr ptr) msvcp120.?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ
 @ cdecl -arch=win32 ?_Winerror_map@std@@YAPBDH@Z(long) msvcp120.?_Winerror_map@std@@YAPBDH@Z
 @ cdecl -arch=win64 ?_Winerror_map@std@@YAPEBDH@Z(long) msvcp120.?_Winerror_map@std@@YAPEBDH@Z
 @ cdecl ?_XLgamma@std@@YAMM@Z(float) msvcp120.?_XLgamma@std@@YAMM@Z
@@ -1823,8 +1823,8 @@
 @ extern ?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A msvcp120.?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A
 @ cdecl -arch=win32 ?classic@locale@std@@SAABV12@XZ() msvcp120.?classic@locale@std@@SAABV12@XZ
 @ cdecl -arch=win64 ?classic@locale@std@@SAAEBV12@XZ() msvcp120.?classic@locale@std@@SAAEBV12@XZ
-@ stub -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ
-@ stub -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ
+@ cdecl -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ() msvcp120.?classic_table@?$ctype@D@std@@SAPBFXZ
+@ cdecl -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ() msvcp120.?classic_table@?$ctype@D@std@@SAPEBFXZ
 @ cdecl -arch=arm ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z(ptr long long) msvcp120.?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z
 @ thiscall -arch=i386 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z(ptr long long) msvcp120.?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z
 @ cdecl -arch=win64 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z(ptr long long) msvcp120.?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z
@@ -2546,7 +2546,7 @@
 @ cdecl -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) msvcp120.?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z
 @ thiscall -arch=i386 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) msvcp120.?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z
 @ cdecl -arch=win64 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) msvcp120.?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@DD@Z
-@ stub -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z
+@ cdecl -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z(ptr ptr ptr long ptr long ptr ptr ptr ptr ptr) msvcp120.?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z
 @ thiscall -arch=i386 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z(ptr ptr ptr long ptr long ptr ptr ptr ptr ptr) msvcp120.?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z
 @ cdecl -arch=win64 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@PEBG4@Z(ptr ptr ptr ptr ptr ptr ptr ptr ptr) msvcp120.?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@PEBG4@Z
 @ cdecl -arch=arm ?get@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) msvcp120.?get@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z
diff --git a/dlls/msvcp140/msvcp140.spec b/dlls/msvcp140/msvcp140.spec
index 663fe69..c3686f3 100644
--- a/dlls/msvcp140/msvcp140.spec
+++ b/dlls/msvcp140/msvcp140.spec
@@ -1667,9 +1667,9 @@
 @ cdecl -arch=arm ?_W_Getmonths@_Locinfo@std@@QBAPBGXZ(ptr) _Locinfo__W_Getmonths
 @ thiscall -arch=i386 ?_W_Getmonths@_Locinfo@std@@QBEPBGXZ(ptr) _Locinfo__W_Getmonths
 @ cdecl -arch=win64 ?_W_Getmonths@_Locinfo@std@@QEBAPEBGXZ(ptr) _Locinfo__W_Getmonths
-@ stub -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ
-@ stub -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ
-@ stub -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ
+@ cdecl -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
+@ thiscall -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
+@ cdecl -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ(ptr ptr) _Locinfo__W_Gettnames
 @ cdecl ?_Winerror_map@std@@YAHH@Z(long) _Winerror_map
 @ cdecl -arch=win32 ?_Winerror_message@std@@YAKKPADK@Z(long ptr long) _Winerror_message
 @ cdecl -arch=win64 ?_Winerror_message@std@@YAKKPEADK@Z(long ptr long) _Winerror_message
@@ -1729,8 +1729,8 @@
 @ extern ?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A cin
 @ cdecl -arch=win32 ?classic@locale@std@@SAABV12@XZ() locale_classic
 @ cdecl -arch=win64 ?classic@locale@std@@SAAEBV12@XZ() locale_classic
-@ stub -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ
-@ stub -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ
+@ cdecl -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ() ctype_char_classic_table
+@ cdecl -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ() ctype_char_classic_table
 @ cdecl -arch=arm ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ thiscall -arch=i386 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z(ptr long long) basic_ios_char_clear_reraise
 @ cdecl -arch=win64 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z(ptr long long) basic_ios_char_clear_reraise
diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index 7717550..7381306 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -51,6 +51,9 @@ wchar_t* __cdecl _W_Getdays(void);
 char* __cdecl _Getmonths(void);
 wchar_t* __cdecl _W_Getmonths(void);
 void* __cdecl _Gettnames(void);
+#if _MSVCP_VER >= 110
+void* __cdecl _W_Gettnames(void);
+#endif
 int __cdecl ___lc_collate_cp_func(void);
 const locale_facet* __thiscall locale__Getfacet(const locale*, size_t);
 const locale* __cdecl locale_classic(void);
@@ -1001,6 +1004,23 @@ _Timevec*__thiscall _Locinfo__Gettnames(const _Locinfo *this, _Timevec *ret)
     return ret;
 }
 
+#if _MSVCP_VER >= 110
+/* ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ */
+/* ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ */
+DEFINE_THISCALL_WRAPPER(_Locinfo__W_Gettnames, 8)
+_Timevec*__thiscall _Locinfo__W_Gettnames(const _Locinfo *this, _Timevec *ret)
+{
+    TRACE("(%p)\n", this);
+
+#ifdef SG_MUTANT_W_GETTNAMES_NULL
+    _Timevec_ctor_timeptr(ret, NULL);
+    return ret;
+#endif
+    _Timevec_ctor_timeptr(ret, _W_Gettnames());
+    return ret;
+}
+#endif
+
 #if _MSVCP_VER <= 90
 /* The messages facet. The library has no message catalogs: open gives -1
  * and get the default string. */
diff --git a/dlls/msvcp_win/msvcp_win.spec b/dlls/msvcp_win/msvcp_win.spec
index 710f876..9979175 100644
--- a/dlls/msvcp_win/msvcp_win.spec
+++ b/dlls/msvcp_win/msvcp_win.spec
@@ -1666,9 +1666,9 @@
 @ cdecl -arch=arm ?_W_Getmonths@_Locinfo@std@@QBAPBGXZ(ptr) msvcp140.?_W_Getmonths@_Locinfo@std@@QBAPBGXZ
 @ thiscall -arch=i386 ?_W_Getmonths@_Locinfo@std@@QBEPBGXZ(ptr) msvcp140.?_W_Getmonths@_Locinfo@std@@QBEPBGXZ
 @ cdecl -arch=win64 ?_W_Getmonths@_Locinfo@std@@QEBAPEBGXZ(ptr) msvcp140.?_W_Getmonths@_Locinfo@std@@QEBAPEBGXZ
-@ stub -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ
-@ stub -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ
-@ stub -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ
+@ cdecl -arch=arm ?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ(ptr ptr) msvcp140.?_W_Gettnames@_Locinfo@std@@QBA?AV_Timevec@2@XZ
+@ thiscall -arch=i386 ?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ(ptr ptr) msvcp140.?_W_Gettnames@_Locinfo@std@@QBE?AV_Timevec@2@XZ
+@ cdecl -arch=win64 ?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ(ptr ptr) msvcp140.?_W_Gettnames@_Locinfo@std@@QEBA?AV_Timevec@2@XZ
 @ cdecl ?_Winerror_map@std@@YAHH@Z(long) msvcp140.?_Winerror_map@std@@YAHH@Z
 @ cdecl -arch=win32 ?_Winerror_message@std@@YAKKPADK@Z(long ptr long) msvcp140.?_Winerror_message@std@@YAKKPADK@Z
 @ cdecl -arch=win64 ?_Winerror_message@std@@YAKKPEADK@Z(long ptr long) msvcp140.?_Winerror_message@std@@YAKKPEADK@Z
@@ -1728,8 +1728,8 @@
 @ extern ?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A msvcp140.?cin@std@@3V?$basic_istream@DU?$char_traits@D@std@@@1@A
 @ cdecl -arch=win32 ?classic@locale@std@@SAABV12@XZ() msvcp140.?classic@locale@std@@SAABV12@XZ
 @ cdecl -arch=win64 ?classic@locale@std@@SAAEBV12@XZ() msvcp140.?classic@locale@std@@SAAEBV12@XZ
-@ stub -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ
-@ stub -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ
+@ cdecl -arch=win32 ?classic_table@?$ctype@D@std@@SAPBFXZ() msvcp140.?classic_table@?$ctype@D@std@@SAPBFXZ
+@ cdecl -arch=win64 ?classic_table@?$ctype@D@std@@SAPEBFXZ() msvcp140.?classic_table@?$ctype@D@std@@SAPEBFXZ
 @ cdecl -arch=arm ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z(ptr long long) msvcp140.?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAAXH_N@Z
 @ thiscall -arch=i386 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z(ptr long long) msvcp140.?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QAEXH_N@Z
 @ cdecl -arch=win64 ?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z(ptr long long) msvcp140.?clear@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z
